more visibility to setting <LangVersion>

BillWagner opened the pull request:
Fixes #2821 

This PR creates a new article under "What's New in C#" for configuring the language version.  That article lists all the versions, and describes the common methods for setting it.

There were a number of locations in the C# Guide that referenced setting the `LangVersion` element. Those have been consolidated in the new article, and the previous locations point to the new article.

I've marked it as WIP because I have two remaining tasks and questions:

- [x] The msbuild / CLI documentation does not reference the `LangVersion` element. It should. @mairaw which article would you recommend updating?
- [x] Parallel changes and a new article should be made in the VB guide. @rpetrusha if you like this direction, I'll make the changes in the VB Guide as well.

BillWagner committed:
more visibility to setting <LangVersion>

This PR creates a new article under "What's New in C#" for configuring the language version.  That article lists all the versions, and describes the common methods for setting it.

dasMulli commented:
You could also include a section describing how to set it for a complete solution at once using a `Directory.Bulid.props` file.

BillWagner committed:
fix build errors

rpetrusha commented:
In terms of the LangVersion in the VB docs, I completely agree that adding it would be great. If you want me to do it, let me know, @BillWagner.

mairaw commented:
Regarding documenting the LangVersion element, we don't have that topic yet @BillWagner. It has been forever in my backlog to create the csproj reference (issue #2642)/

BillWagner committed:
Add discussion of Directory.build.props

BillWagner committed:
add VB version article

Also, provide links from locations where the version was specified.

BillWagner committed:
add note to include

Trying a test build

BillWagner committed:
fix build errors

BillWagner commented:
Thanks for the answers.
@dasMulli I added as section on using the Directory.build.props file.
@rpetrusha The last commits have the VB versions
@mairaw I added a comment on #2642 to note these new files.

rpetrusha commented on docs/csharp/whats-new/configure-language-version.md:
```diff
@@ -0,0 +1,92 @@
+
+## Set the language version in Visual Studio
+
+You can set the version in Visual Studio. Right-click on the project node in Solution Explorer and select **Properties**. Select the **Build** tab and select the **Advanced** button. In the dropdown, select the version. The following image shows the "latest" setting.
```
There isn't an image.

rpetrusha commented on docs/csharp/whats-new/configure-language-version.md:
```diff
@@ -0,0 +1,92 @@
+> creates separate nodes for each build configuration. You'll typically
+> set the value the same in all build configurations, but you need to
+> set it explicitly for each build configuration, or select "All Configurations"
+> when you modify this setting. You'll see the following in your csproj file:
```
nit: It's somewhat odd to refer to "the following" in a note box, and not have the code included in the note.

rpetrusha commented on docs/csharp/whats-new/configure-language-version.md:
```diff
@@ -0,0 +1,92 @@
+
+## Edit the csproj file
+
+You can set the language version in your **.csproj** file. Add the following element:
```
instead of "the following", "an element with a value like the following" perhaps

rpetrusha commented on docs/csharp/whats-new/configure-language-version.md:
```diff
@@ -0,0 +1,92 @@
+|7.3|The compiler accepts only syntax that is included in C# 7.3 or lower.|
+|latest|The compiler accepts all valid language syntax that it can support.|
+
+The special strings `default` and `latest` resolve to the latest major
```
It might be good to give examples of latest major (C# 7.0) and minor versions.

rpetrusha commented on docs/toc.md:
```diff
@@ -215,6 +215,7 @@
 <!-- The "What's New" section is short, and one level
     deep, so leave it in the main TOC -->
 ## [What's new in C#](csharp/whats-new/index.md)
+### [Configure language version](csharp/whats-new/configure-language-version.md)
```
I wouldn't expect to find these topics in a what's new section; they're more reference-like -- possibly as immediate children of the Language Reference topic?

rpetrusha commented on docs/visual-basic/getting-started/configure-language-version.md:
```diff
@@ -0,0 +1,63 @@
+There are three ways to set the language version:
+
+- Manually edit your [**.vbproj** file](#edit-the-vbproj-file)
+- Set the language version [for multiople projects in a subdirectory](#configure-multiple-projects)
```
multiople --> multiple

rpetrusha commented on docs/visual-basic/getting-started/configure-language-version.md:
```diff
@@ -0,0 +1,63 @@
+|15.5|The compiler accepts only syntax that is included in Visual Basic 15.5 or lower.|
+|latest|The compiler accepts all valid language syntax that it can support.|
+
+The special strings `default` and `latest` resolve to the latest major
```
give an example, just in case someone doesn't know what major and minor are.

rpetrusha commented on docs/visual-basic/getting-started/configure-language-version.md:
````diff
@@ -0,0 +1,63 @@
+</Project>
+```
+
+Now, builds in every sub-directory of the directory containing that file will use Visual Basic version 15.5 syntax. For more information, see the article on [Customize your build](/visualstudio/msbuild/customize-your-build.md).
````
I think that subdirectory is one word.

rpetrusha commented on docs/visual-basic/getting-started/configure-language-version.md:
```diff
@@ -0,0 +1,63 @@
+
+## Set the langversion compiler option
+
+You can use the `-langversion` command line option. For more information, see the article on the [--langversion](../reference/command-line-compiler/langversion.md) compiler option.
```
just `-langversion`

rpetrusha commented on docs/csharp/whats-new/configure-language-version.md:
```diff
@@ -0,0 +1,92 @@
+
+## Set the langversion compiler option
+
+You can use the `-langversion` command line option. For more information, see the article on the [--langversion](../language-reference/compiler-options/langversion-compiler-option.md) compiler option.
```
just `-langversion`

rpetrusha commented on docs/visual-basic/reference/command-line-compiler/langversion.md:
```diff
@@ -18,7 +18,11 @@ Causes the compiler to accept only syntax that is included in the specified Visu
+ Any of the whole numbers may also be specified using `.0` as the minor version, for example, `11.0`.
+
+ You can see the list of all possible values by specifying `-langversion:?` on the command line.  
   
```
For both the C# and VB topics, the syntax to display a list of language versions (`-langversion:?`) should be mentioned.

rpetrusha reviewed (changes requested):
This is a really valuable addition to the documentation, @BillWagner! I've left a number of comments for you to address. I'm also not certain that the what's new section is the best place for these topics -- I'd expect to find them in the language reference.

BillWagner committed:
respond to feedback

I had to rebase for conflicts before this commit.

To see just the latest changes, look at this commit only.

BillWagner commented:
@rpetrusha I updated this based on your feedback. It should be ready for another review.

mairaw commented on docs/visual-basic/language-reference/modifiers/private-protected.md:
````diff
@@ -19,6 +19,8 @@ The `Private Protected` access modifier is supported by Visual Basic 15.5 and la
 </PropertyGroup>
 ```
 
+For more information see [setting the Visual Basic language version](../../language-reference/configure-language-version.md).
````
can you escape the * on line 14 so this doesn't look italic here?

mairaw commented on docs/visual-basic/programming-guide/language-features/data-types/tuples.md:
```diff
@@ -56,6 +56,8 @@ To enable interred tuple element names, you must define the version of the Visua
 
 The version number can be any version of the Visual Basic compiler starting with 15.3. Rather than hard-coding a specific compiler version, you can also specify "Latest" as the value of `LangVersion` to compile with the most recent version of the Visual Basic compiler installed on your system.
 
+For more information see [setting the Visual Basic language version](../../../language-reference/configure-language-version.md).
```
missing comma on these sentence after For more information

mairaw commented on docs/visual-basic/language-reference/configure-language-version.md:
```diff
@@ -0,0 +1,63 @@
+---
+title: Select the Visual Basic language version
+description: Configure the compiler to perform syntax validation using a specific compiler version
```
missing punctuation

mairaw commented on docs/visual-basic/language-reference/configure-language-version.md:
```diff
@@ -0,0 +1,63 @@
+title: Select the Visual Basic language version
+description: Configure the compiler to perform syntax validation using a specific compiler version
+ms.date: 05/24/2018
+---
```
add author and ms.author here since we don't have defaults for VB

mairaw commented:
Build seems stuck. I'll close and reopen this PR.

mairaw commented on docs/csharp/language-reference/configure-language-version.md:
```diff
@@ -0,0 +1,95 @@
+
+## Set the language version in Visual Studio
+
+You can set the version in Visual Studio. Right-click on the project node in Solution Explorer and select **Properties**. Select the **Build** tab and select the **Advanced** button. In the dropdown, select the version. The following image shows the "latest" setting.
```
nit: end with :

mairaw commented on docs/csharp/language-reference/configure-language-version.md:
```diff
@@ -0,0 +1,95 @@
+
+You can set the version in Visual Studio. Right-click on the project node in Solution Explorer and select **Properties**. Select the **Build** tab and select the **Advanced** button. In the dropdown, select the version. The following image shows the "latest" setting.
+
+![Setting the language version](./media/configure-language-version/advanced-build-settings.png)
```
more info on this alt text? perhaps Screenshot of...

mairaw commented on docs/csharp/language-reference/configure-language-version.md:
```diff
@@ -0,0 +1,95 @@
+
+## Set the langversion compiler option
+
+You can use the `-langversion` command line option. For more information, see the article on the [-langversion](../language-reference/compiler-options/langversion-compiler-option.md) compiler option. You can see a list of the valid values by typing  `vbc -langversion:?`.
```
command line option -> command-line option

vbc? shouldn't be csc here? should you tell that they need to type on the developer command prompt for vs?

BillWagner replied:
and, updated command-line in the VB specific doc.

mairaw reviewed (commented):
Looks great @BillWagner. Left a  few comments.

BillWagner committed:
respond to feedback.

rpetrusha reviewed (approved):
This looks great, @BillWagner. It's ready to merge.
